oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/MetadataRegistry.java

```diff
@@ -0,0 +1,433 @@
+
+        private List<TagFamilySpec> extractTagFamilySpec(List<TagMetadata> tagMetadataList) {
+            Map<String, List<TagMetadata>> tagMetadataMap = tagMetadataList.stream()
+                    .collect(Collectors.groupingBy(tagMetadata -> tagMetadata.isIndex() ? PartialMetadata.this.indexFamily() : PartialMetadata.this.nonIndexFamily()));
+
+            final List<TagFamilySpec> tagFamilySpecs = new ArrayList<>(tagMetadataMap.size());
+            for (final Map.Entry<String, List<TagMetadata>> entry : tagMetadataMap.entrySet()) {
+                final TagFamilySpec.Builder b = TagFamilySpec.create(entry.getKey())
+                        .addTagSpecs(entry.getValue().stream().map(TagMetadata::getTagSpec).collect(Collectors.toList()));
+                if (this.getKind() == Kind.MEASURE && entry.getKey().equals(this.indexFamily())) {
+                    // append measure ID, but it should not generate an index in the client side.
+                    // BanyanDB will take care of the ID index registration.
+                    b.addIDTagSpec();
+                }
+                tagFamilySpecs.add(b.build());
+            }
+
+            return tagFamilySpecs;
+        }
+
+        public Group getOrCreateGroup(BanyanDBClient client) throws BanyanDBException {
+            Group g = client.findGroup(this.group);
+            if (g != null) {
+                return g;
+            }
+            switch (kind) {
+                case STREAM:
+                    return client.define(Group.create(this.group, Catalog.STREAM, 2, 0, Duration.ofDays(7)));
+                case MEASURE:
+                    return client.define(Group.create(this.group, Catalog.MEASURE, 2, 12, Duration.ofDays(7)));
+                default:
+                    throw new IllegalStateException("should not reach here");
+            }
+        }
+
+        public String indexFamily() {
+            switch (kind) {
+                case MEASURE:
+                    return "default";
+                case STREAM:
+                    return "searchable";
+                default:
+                    throw new IllegalStateException("should not reach here");
+            }
+        }
+
+        public String nonIndexFamily() {
+            switch (kind) {
+                case MEASURE:
+                case STREAM:
+                    return "binary";
+                default:
+                    throw new IllegalStateException("should not reach here");
+            }
+        }
+    }
+
+    public enum Kind {
+        MEASURE, STREAM;
+    }
+
+    @RequiredArgsConstructor
+    @Getter
+    private static class TagMetadata {
+        private final IndexRule indexRule;
+        private final TagFamilySpec.TagSpec tagSpec;
+
+        boolean isIndex() {
+            return this.indexRule != null;
+        }
+    }
+
+    @Builder
+    @EqualsAndHashCode
+    public static class Schema {
+        @Getter
+        private final PartialMetadata metadata;
```

Comment:
Same question about this class name `PartialMetadata`.
As an official schema, it should not include a thing called `partial*`.

Reply:
Renamed to `SchemaMetadata`